feat: Status Page Custom CSS/JS/HTML overrides (Fixes #2863)

Armaansaxena opened the pull request:
## Describe your changes
I have implemented the ability for admins to customize their Status Pages with custom CSS, JavaScript, and HTML overrides for the Header and Footer. This allows teams to brand their status pages without forking the source code.

### Backend Changes:
- Updated `StatusPage` Mongoose schema to include `customJavaScript`, `headerHTML`, and `footerHTML` (added to existing `customCSS`).
- Updated Joi validation (`server/validation/joi.js`) to allow these string fields during creation and updates.

### Frontend Changes:
- Updated `Settings.jsx` to include input fields for these new options, ensuring correct name attributes for form saving.
- Added translation keys to `en.json` to ensure no hardcoded strings are used in the UI.
- Updated `Settings.jsx` to inject the custom CSS/JS via `useEffect` and conditionally render the custom HTML Header/Footer if they exist in the database.

## Write your issue number after "Fixes "
**Fixes #2863**

## Please ensure all items are checked off before requesting a review. "Checked off" means you need to add an "x" character between brackets so they turn into checkmarks.

- [x] (Do not skip this or your PR will be closed) I deployed the application locally.
- [x] (Do not skip this or your PR will be closed) I have performed a self-review and testing of my code.
- [x] I have included the issue # in the PR.
- [x] I have added i18n support to visible strings (instead of `<div>Add</div>`, use):
```javascript
const { t } = useTranslation();
<div>{t('add')}</div>

<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

* **New Features**
  * Add custom CSS, header HTML, footer HTML, and custom JavaScript for Status Pages; header/footer and assets are rendered on the public page with proper injection and cleanup.
  * Custom JavaScript is disabled until users explicitly accept the risk.

* **Security**
  * User-supplied HTML/JS is sanitized before rendering.

* **Documentation**
  * Added UI labels, descriptions, placeholders and a JavaScript warning helper.

* **Validation**
  * Server validates the new optional fields with length limits.

<sub>✏️ Tip: You can customize this high-level summary in your review settings.</sub>
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

Armaansaxena committed:
feat: Status Page Custom CSS/JS/HTML overrides (Fixes #2863)

Armaansaxena committed:
Update server/src/validation/joi.js

Co-authored-by: llamapreview[bot] <184758061+llamapreview[bot]@users.noreply.github.com>

SajanGhuman commented on client/src/Pages/v1/StatusPage/Status/index.jsx:
```diff
@@ -16,151 +16,204 @@ import { useIsAdmin } from "../../../../Hooks/v1/useIsAdmin.js";
+                script.id = 'custom-status-js';
+                // Wrap in IIFE to avoid global scope pollution
+                script.text = `(function() { ${statusPage.customJavaScript} })();`;
+                document.body.appendChild(script);
```
@Armaansaxena can you explain, why you are injecting customJS on component mount and using appendChld? I understand that intention is custom branding but we are essentially allowing anybody in control of the status file to run custom scripts on all user's browser.

@ajhollid

Armaansaxena replied:
@SajanGhuman Valid question.

1. The "Why" (Use Case): The primary requirement for customJavaScript is to allow Admins to integrate third-party tools like Google Analytics, Intercom, or Chat Widgets on their status pages. These integrations universally require a script tag to be injected into the DOM to fetch their external resources.

2. The Implementation: I am using appendChild because React's dangerouslySetInnerHTML intentionally does not execute <script> tags for security reasons. To actually load a chat widget or analytics tracker, manual DOM manipulation via document.createElement is necessary.

Addressing the Risk: I fully agree this opens a Stored XSS vector if an Admin account is compromised. To mitigate this:

I am adding a prominent "Security Warning" on the settings page next to this input.

I am implementing DOMPurify for the HTML fields (Header/Footer) to strictly sanitize non-script inputs.

If you believe the risk of allowing JS overrides outweighs the benefit of having Analytics/Chat support, I am happy to remove the JS field entirely and stick to just CSS and Sanitized HTML. Please let me know your preference.

SajanGhuman commented:
The headerHTML and footerHTML fields introduce raw HTML injection into the page. As CodeRabbit is warning, this is essentially creating an critical XSS risk. I understand that the feature is admin-only but rendering untrusted HTML and directly injecting JS is a bad practice overall.

Armaansaxena commented:
@SajanGhuman  @ajhollid 
Thank you for the critical review. I fully agree with the security concerns raised regarding the XSS risks and arbitrary code execution.

My plan to fix this:

HTML Sanitization: I will implement DOMPurify to sanitize both headerHTML and footerHTML before rendering, ensuring no malicious scripts can be injected via these fields.

Backend Limits: I will update the Joi validation to enforce strict character limits (as suggested by CodeRabbit) to prevent database performance issues.

Custom JS: I understand the risk here. Since this was a requested feature in issue #2863 (allowing admins to override behavior), I will add a strict "Risk Acceptance" warning in the UI that the admin must check before saving.

If you believe customJS is too dangerous even with admin warnings, please let me know, and I can remove that specific field and limit the scope to just CSS and Sanitized HTML.

I will push these fixes shortly.

Armaansaxena committed:
fix: address security review - safe useEffect cleanup, dompurify sanitization, and validation limits

Armaansaxena committed:
Merge remote changes and fix validation limits

ajhollid commented:
Hi @Armaansaxena,

Perhaps the solution is to gate this feature behind some sort of approval process whereby the admin acknowledges the risks of allowing arbitrary code to be executed.

@SajanGhuman @gorkemcetin what do you think?

Armaansaxena committed:
fix: apply critical security fixes (sanitization, props, schema limits)

SajanGhuman commented:
> Hi @Armaansaxena,
> 
> Perhaps the solution is to gate this feature behind some sort of approval process whereby the admin acknowledges the risks of allowing arbitrary code to be executed.
> 
> @SajanGhuman @gorkemcetin what do you think?

I agree. There should be some kind of warning explaining the risks that the admin is aware of and acknowledges to.

Armaansaxena committed:
fix: sanitize custom CSS and enforce JS risk acceptance

Armaansaxena committed:
fix: harden css sanitization regex per coderabbit review

gorkem-bwl commented:
> > Hi @Armaansaxena,
> > Perhaps the solution is to gate this feature behind some sort of approval process whereby the admin acknowledges the risks of allowing arbitrary code to be executed.
> > @SajanGhuman @gorkemcetin what do you think?
> 
> I agree. There should be some kind of warning explaining the risks that the admin is aware of and acknowledges to.

Just a warning should be fine.

Armaansaxena committed:
fix: correct i18n fallback syntax for security warning

Armaansaxena commented:
@gorkem-bwl @SajanGhuman @ajhollid

**Update:** All requested security controls and fixes are implemented.

Based on the feedback, I have finalized the security model for this feature:

- JS Risk Gatekeeper (Implemented): The "Custom JavaScript" input is now disabled by default. The Admin must strictly check a "Risk Acceptance" checkbox to unlock the field.

- CSS Sanitization (Hardened): Per CodeRabbit's feedback, I updated the CSS injection to strictly block dangerous patterns (e.g., url(), @import, behavior, javascript:).

- HTML Sanitization: Full DOMPurify implementation for Header/Footer is in place.

- Backend Validation: Added strict maxLength constraints to the schema.

- Fixes: Resolved the i18n fallback syntax and variable naming conflicts.

The feature is now secure, gated, and validates all inputs. Ready for merge!

Armaansaxena committed:
chore: resolve restructuring conflicts and synchronize with upstream develop
